Expand `condition_script_runner_args`

wmmc88 opened the pull request:
#1081 added `condition_script_runner_args`, but `condition_script_runner_args` didn't expand env vars like `script_runner_args` did. This PR adds env expansion to `condition_script_runner_args`, and also adds tests for it.

Various bugfixes to tests were also fixed in order for `cargo make ci-flow` to pass on:
1. windows computers
2. environments that set `CARGO_HOME`

wmmc88 commented on src/lib/runner.rs:
```diff
@@ -361,6 +361,28 @@ pub(crate) fn run_task_with_options(
             None => (),
         };
 
+        let step = {
```
this block needed to be moved up so that expansion happens before condition validation (since condition validation calls the script runner with the args)

sagiegurari replied:
i get it, but its not really what you want. you just want to expand.
this entire block also updated the current env.
so in case condition fails, the env is still updated. this is unexpected.

wmmc88 replied:
Ok. Let me fix that.


For the evaluation of the condition script, is the correct behaviour to not use any env information defined in the current task when evaluating the condition for the current task?

sagiegurari commented:
@wmmc88 thanks for the PR! would love if you see my comment

wmmc88 commented on README.md:
```diff
@@ -1652,16 +1664,16 @@ These scripts use that value to create a new environment variable **`COMPOSITE_2
 * Load global environment variables defined in the **env.\[current profile\]** block.
 * Load global environment setup scripts defined in the **env_scripts** attribute.
 * **Per Task**
-  * Load environment files defined in the **env_files** attribute (relative paths are treated differently than global env_files).
   * Setup **per task** internal environment variables (see [Global](#usage-env-global) section).
+  * Load environment files defined in the **env_files** attribute (relative paths are treated differently than global env_files).
   * Load environment variables defined in the **env** block (same behavior as global env block).
```
@sagiegurari This looks like the existing documentation was wrong, so I've corrected it to match what the code actually does (i.e. internal are expanded before env_files

wmmc88 commented:
> @wmmc88 thanks for the PR! would love if you see my comment

@sagiegurari I've updated the PR so that there is no change in env when condition evaluation fails (and added a test to prevent possible future regressions)

wmmc88 commented:
@sagiegurari friendly ping to review this :) the underlying bug is a blocker for me

sagiegurari commented:
@wmmc88 sorry for the delay! i'll go over it today

sagiegurari commented on README.md:
```diff
@@ -10,132 +10,144 @@
 
 > [Rust](https://www.rust-lang.org/) task runner and build tool.
 
-* [Overview](#overview)
```
all changes in this file will be deleted as its auto generated.
nav changes go to nav.md (but please do'nt add that extra identation, not sure we need that new top level cargo-make item as we are already at the top)
and content changes go to content.md
both files are under docs/_includes/

wmmc88 replied:
I did not add nav changes intentionally. I think I just ran `cargo make ci-flow` locally, and this is what happened. I've reverted these changes and re-added the content changes to `content.md`.

sagiegurari commented on src/lib/runner.rs:
```diff
@@ -465,6 +466,11 @@ pub(crate) fn run_task_with_options(
             }
         } else {
+            // Restore env to state before expansion for condition evaluation
+            for (key, value) in env_before_current_task_env_expansion {
+                envmnt::set(key, value);
+            }
```
the issue here is that new env vars are still stuck in the env. you only restore modified env vars.

sagiegurari replied:
@wmmc88 rethinking this whole thing, can we just not do all this env changes and only expand with what we have?
it is unexpected bahvior that env changes before condition is met.
if the condition needs some env vars, let them be defined in a dependency task as is.
this is much more readable and expected

wmmc88 replied:
I could rework the pr to not expand the env for the condition script, but I want to highlight that this means the some env will change before condition script and some will change after. Particularly, all the internal meta ends like `CARGO_MAKE_CURRENT_TASK*` will expand before condition, and the rest will expand after.

sagiegurari replied:
i think what i'm saying is that for your PR, runner.rs does not need to change at all and control should be in the makefile author to setup env for conditions before hand

wmmc88 replied:
Before condition evaluation in runner.rs, there still needs to be a small change to call a function to expand args for condition_script_runner_args. 

If I understand you correctly, the `CARGO_MAKE_CURRENT_TASK_*` env vars will expand to the updated values for the current task (which matches existing behavior), but other env vars will not be updated (i.e. I remove the env expansion).

sagiegurari replied:
if its just to call expand ya thats ok.

env is updated for all vars. no env var gets special invocation

wmmc88 replied:
made change to only explicitly expand the condition-script-runner-args

sagiegurari reviewed (changes requested):
@wmmc88 everything looks pretty great just 2 main thinigs
1. docs - the file is generated so your changes will be deleted. lets try not to do the nav changes you did but only the content. you can add nav, but not add a new root please.
2. env setup before the condition and restoring it causes new env vars defined in the task to stick even if condition fails. we need to remove them.

wmmc88 commented:
@sagiegurari I think I've resolved all your requested changes

sagiegurari commented:
Thanks a lot @wmmc88 for the PR. merged.
I'll probably release a new version at end of week.
sorry it took so long.

wmmc88 commented:
> Thanks a lot @wmmc88 for the PR. merged. I'll probably release a new version at end of week. sorry it took so long.

Hi @sagiegurari! friendly bump for a release
